Python 3.10 support

tnoczyns-volue opened the pull request:
~~Put on hold - due to current incompatibility of pip/distutil (distutil is needed for kerberos on Ubuntu) with Python 3.10.~~

~~For windows github actions are working correctly for python 3.10.~~

We are supporting Python 3.10 now.

torsannes commented:
@tnoczyns-volue - we got a request from Agder to be on Python 3.10 (or 3.11). I see in this PR that there are an issue on 3.10 when running on Ubuntu, is that possible to fix or to ignore (and make a statement in documentation)?

tnoczyns-volue commented:
I'd need to recheck. Meanwhile also Python 3.11 was released, so I can check this as well. I'll get back with answer.

aes-powel commented:
Will/do we have an official strategy for dealing with new Python releases? i.e commit to supporting the x to last available version within y weeks after a new release?

tnoczyns-volue commented:
Currently latest version of `pyarrow` (10.0.0 released in Oct 27th, 2022) is not supporting Python 3.11.
PyArrow is one of Mesh Python SDK dependencies.

tnoczyns-volue commented:
> Will/do we have an official strategy for dealing with new Python releases? i.e commit to supporting the x to last available version within y weeks after a new release? This will not be a one time issue.

I'm not aware of such plan and I'm not sure what was the agreement.
I'll just share my view: supporting Python versions means that:
* our own code is compatible
* our dependencies (meaning other packages that we rely on) are compatible
 
While generally I don't see a big problem with changing our code to support newest Python versions, the problem may come from our dependencies. For example `pyarrow` is currently not supporting Python 3.11. In such case we need to wait until the owner of such component makes a new, compatible release. I'd like to avoid the need of contributing to 3rd party components to make them support newer Python version faster.

That is why I'd refrain of making such commitment. We are still supporting 3 currently maintained Python versions:
https://devguide.python.org/versions/#supported-versions

On the other hand I also understand clients wanting to use newest Python versions, due to e.g. performance improvements.

tnoczyns-volue commented:
FYI: since we have some discussion here, I forced pushed changes from https://github.com/Volue-Public/energy-mesh-python/pull/334.

tnoczyns-volue commented on pyproject.toml:
```diff
@@ -19,9 +19,9 @@ include = [
 
 [tool.poetry.dependencies]
-python = "^3.7.1" #pandas require python 3.7.1 or above
+python = ">=3.7.1,<3.11"  # pandas requires python 3.7.1 or above, numpy (one of the dependencies) requires python below 3.11
```
This is ugly, but with the next version we'll just do `>=3.8`

tnoczyns-volue commented on docs/source/versions.rst:
```diff
@@ -3,6 +3,36 @@ Versions
+- Support for Python 3.10 :pull:`93`
+
+.. warning::
+    Python 3.7.1 support will dropped in the next Mesh Python SDK release.
```
With new Python 3.11 and Python 3.7.1 end of life in couple of months (June 2023) I think we can already warn the users about dropping this in the next release. Support for 3.7 complicates things with out dependencies, so I prefer to go from 3.8 and try supporting 3.11 when possible.

torsannes replied:
Agreed - we can probably have a "session" every 6 month including new Python versions and dropping older Python version that is close to end-of-life (within the next 6 months). This strategy should be documented in the version description.

tnoczyns-volue commented:
There were couple of problems:
* Problems with resolving dependencies versions by poetry.
* Pip for Python 3.10: Ubuntu 20.04 installs pip for Python 3.8. For Python 3.10 we need the latest version. We can e.g. download it directly from `https://bootstrap.pypa.io/get-pip.py`. When using `venv` more up-to-date pip version should be already used.
* `kerberos` wheel wasn't building for Python 3.10 and required `gcc` to be installed.

I captured those points in FAQ section in our documentation, updated dependencies in `pyproject.toml`, added check for 3.10 in GitHub Actions and created entry on new version in documentation. I'll create a new 1.1.0 release after this is merged.

erny-powel reviewed (approved):
Looks great
